openedx/core/djangoapps/waffle_utils/views.py

```diff
@@ -3,277 +3,124 @@
+    )
+    permission_classes = (
+        permissions.IsAuthenticated,
+        IsStaff,
+    )
 
-def _add_waffle_switch_computed_status(switch_dict):
-    """
-    Add computed status to each waffle switch.
-    """
-    for switch in switch_dict.values():
-        computed_status = 'off'
-        if 'is_active' in switch:
-            if switch['is_active'] == 'true':
-                computed_status = 'on'
-            else:
-                computed_status = 'off'
-        switch['computed_status'] = computed_status
-
+    def get(self, request):
+        """
+        Expose toggle state report dict as a view.
+        """
+        report = CourseOverrideToggleStateReport().as_dict()
+        _add_waffle_flag_course_override_state(report["waffle_flags"])
+        return Response(report)
 
-def _get_all_waffle_flags():
-    """
-    Gets all waffle flags and their state.
-    """
-    flags_dict = {}
-    _add_waffle_flag_instances(flags_dict)
-    _add_waffle_flag_state(flags_dict)
-    _add_waffle_flag_course_override_state(flags_dict)
-    _add_waffle_flag_computed_status(flags_dict)
-    flag_list = list(flags_dict.values())
-    flag_list.sort(key=lambda toggle: toggle['name'])
-    return flag_list
 
-
-def _add_waffle_flag_instances(flags_dict):
+def _add_waffle_flag_course_override_state(flags_dict):
     """
-    Add details from waffle flag instances, like code_owner.
+    Add waffle flag course override state from the WaffleFlagCourseOverrideModel model.
     """
-    waffle_flag_instances = WaffleFlag.get_instances()
-    for flag_instance in waffle_flag_instances:
-        flag = _get_or_create_toggle_response(flags_dict, flag_instance.name)
-        _add_toggle_instance_details(flag, flag_instance)
 
-
-def _add_waffle_flag_state(flags_dict):
-    """
-    Add waffle flag state from the waffle Flag model.
-    """
-    waffle_flags = Flag.objects.all()
-    for flag_data in waffle_flags:
-        flag = _get_or_create_toggle_response(flags_dict, flag_data.name)
-        if flag_data.everyone is True:
-            everyone = 'yes'
-        elif flag_data.everyone is False:
-            everyone = 'no'
-        else:
-            everyone = 'unknown'
-        flag['everyone'] = everyone
-        if flag_data.note:
-            flag['note'] = flag_data.note
-        flag['created'] = str(flag_data.created)
-        flag['modified'] = str(flag_data.modified)
+    flag_course_overrides = _get_flag_course_overrides()
+    for flag_name, course_overrides_dict in flag_course_overrides.items():
+        course_overrides = [
+            course_override
+            for course_override in course_overrides_dict.values()
+            if not course_override.get("disabled")
+        ]
+        if course_overrides:
+            flag = get_or_create_toggle_response(flags_dict, flag_name)
+            flag["course_overrides"] = course_overrides
 
 
-def _add_waffle_flag_course_override_state(flags_dict):
+def _get_flag_course_overrides():
```

Comment:
Should we enhance test coverage? Note: adding the test before the refactor would ensure the refactor isn’t changing anything.

Reply:
Again, test coverage could certainly be improved, yes.

Reply:
Note: The computation in this method is complex and initially had a bug, so I think we want test coverage added.  We added test coverage earlier for this endpoint as part of the blended tasks, but I missed that we didn't cover the course waffle flags.

Could you clarify your plan?  It is unclear if you:
1. Plan to add a test outside of this PR that we could use to ensure this PR doesn't introduce a regression, or
2. Plan to add a test as part of this PR, which would get us the additional test coverage with slightly less confidence, or
3. Agree that coverage could be improved, but don't think it should be prioritized, or
4. Other?

I am proposing we do Plan 1 above. If you disagree, please add some context.  Thank you!

Reply:
I added some tests to this. If we feel like we need to add more test coverage (aka: plan 1) we can do this later.

Reply:
What I did was write a `test_compute_status` in this branch. Then I checked out the master branch and copy-pasted the test there, with only minor modifications. It worked out of the box, so I don't think it's necessary to create another PR.